src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs

```diff
@@ -205,107 +205,79 @@ public class ControllerArgumentBinder : IControllerActionArgumentBinder
             return modelBindingResult;
         }
 
-        // Called via reflection.
-        private static void CallPropertyAddRange<TElement>(object target, object source)
-        {
-            var targetCollection = (ICollection<TElement>)target;
-            var sourceCollection = source as IEnumerable<TElement>;
-            if (sourceCollection != null && !targetCollection.IsReadOnly)
-            {
-                targetCollection.Clear();
-                foreach (var item in sourceCollection)
-                {
-                    targetCollection.Add(item);
-                }
-            }
-        }
-
-        private void ActivateProperties(
-            ControllerActionDescriptor actionDescriptor,
+        private void ActivateProperty(
+            ParameterDescriptor property,
+            PropertyHelper propertyHelper,
             object controller,
-            IDictionary<string, object> properties)
+            object value)
         {
-            var propertyHelpers = PropertyHelper.GetProperties(controller);
-            for (var i = 0; i < actionDescriptor.BoundProperties.Count; i++)
-            {
-                var property = actionDescriptor.BoundProperties[i];
-
-                PropertyHelper propertyHelper = null;
-                for (var j = 0; j < propertyHelpers.Length; j++)
-                {
-                    if (string.Equals(propertyHelpers[j].Name, property.Name, StringComparison.Ordinal))
-                    {
-                        propertyHelper = propertyHelpers[j];
-                        break;
-                    }
-                }
+            var propertyType = propertyHelper.Property.PropertyType;
+            var metadata = _modelMetadataProvider.GetMetadataForType(propertyType);
 
-                object value;
-                if (propertyHelper == null || !properties.TryGetValue(property.Name, out value))
+            if (propertyHelper.Property.CanWrite && propertyHelper.Property.SetMethod?.IsPublic == true)
+            {
+                // Handle settable property. Do not set the property to null if the type is a non-nullable type.
+                if (value != null || metadata.IsReferenceOrNullableType)
                 {
-                    continue;
+                    propertyHelper.SetValue(controller, value);
                 }
```

Comment:
The mean that the special cases that used to be handled by `continue` are now part of the calling code and not part of this method.